src/kernel.rs

```diff
@@ -0,0 +1,100 @@
+//! The kernel type for filter operations
+
+/// An borrowed 2D kernel, used to filter images via convolution.
+#[derive(Debug, Copy, Clone)]
+pub struct Kernel<'a, K> {
+    pub(crate) data: &'a [K],
+    pub(crate) width: u32,
+    pub(crate) height: u32,
+}
+
+impl<'a, K> Kernel<'a, K> {
+    /// Construct a kernel from a slice and its dimensions. The input slice is
+    /// in row-major form.
+    pub fn new(data: &'a [K], width: u32, height: u32) -> Kernel<'a, K> {
+        assert!(width > 0 && height > 0, "width and height must be non-zero");
+        assert!(
+            width * height == data.len() as u32,
+            "Invalid kernel len: expecting {}, found {}",
+            width * height,
+            data.len()
+        );
+        Kernel {
+            data,
+            width,
+            height,
+        }
+    }
+
+    /// Get the value in the kernel at the given `x` and `y` position.
+    #[inline]
+    pub fn get_unchecked(&self, x: u32, y: u32) -> &K {
```

Comment:
`get_unchecked` is typically used for unsafe indexing. I think this is just `get`.

Reply:
To play devil's advocate the theoretical `get_checked` would return an option in case the given indices were out of range rather than panicking which mirrors the various checked Vs unchecked functions in the standard library.

Reply:
To play devil's advocate the theoretical `get_checked` would return an option in case the given indices were out of range rather than panicking which mirrors the various checked Vs unchecked functions in the standard library.

Reply:
In the standard library `get_unchecked` indicates unchecked indexing rather than indexing that might panic, e.g.

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get_unchecked
https://doc.rust-lang.org/std/primitive.str.html#method.get_unchecked

Good point that `get` typically returns an `Option`. Maybe this could just be called `at`?